lambda/tmc.ml

```diff
@@ -0,0 +1,854 @@
+    Llet (Strict, Pgenval, block_var, k_with_placeholder,
+          body { var = block_var; offset = placeholder_pos_lam ; loc = con.loc } (Lvar block_var))
+
+  (** We want to delay the application of the constructor to a later time.
+      This may move the constructor application below some effectful
+      expressions (if we move into a context of the form [foo;
+      bar_with_tmc_inside] for example), and we want to preserve the
+      evaluation order of the other arguments of the constructor.  So we bind
+      them before proceeding, unless they are obviously side-effect free. *)
+  let delay_impure : block_id:int -> t -> (t -> lambda) -> lambda =
+    let bind_list ~block_id ~arg_offset lambdas k =
+      let can_be_delayed =
+        (* Note that the delayed subterms will be used
+           exactly once in the linear-static subterm. So
+           we are happy to delay constants, which we would
+           not want to duplicate. *)
+        function
+        | Lvar _ | Lconst _ -> true
+        | _ -> false in
+      let bindings, args =
+        lambdas
+        |> List.mapi (fun i lam ->
+            if can_be_delayed lam then (None, lam)
+            else begin
+              let v = Ident.create_local
+                  (Printf.sprintf "block%d_arg%d" block_id (arg_offset + i)) in
+              (Some (v, lam), Lvar v)
+            end)
+        |> List.split in
+      let body = k args in
+      List.fold_right (fun binding body ->
+          match binding with
+          | None -> body
+          | Some (v, lam) -> Llet(Strict, Pgenval, v, lam, body)
+        ) bindings body in
+    fun ~block_id con body ->
+    bind_list ~block_id ~arg_offset:0 con.before @@ fun vbefore ->
+    let arg_offset = List.length con.before + 1 in
+    bind_list ~block_id ~arg_offset con.after @@ fun vafter ->
+    body { con with before = vbefore; after = vafter }
+end
+
+(** The type ['a Dps.t] (destination-passing-style) represents a
+    version of ['a] that is parametrized over a [lambda destination].
+    A [lambda Dps.t] is a code fragment in destination-passing-style,
+    a [(lambda * lambda) Dps.t] represents two subterms parametrized
+    over the same destination. *)
+module Dps : sig
+  type 'a dps = tail:bool -> dst:lambda destination -> 'a
+  (** A term parameterized over a destination.  The [tail] argument
+      is passed by the caller to indicate whether the term will be placed
+      in tail-position -- this allows to generate correct @tailcall
+      annotations. *)
+
+  type 'a t
+
+  val make : lambda dps -> lambda t
+  val run : lambda t -> lambda dps
+  val delay_constructor : Constr.t -> lambda t -> lambda t
+
+  val return : lambda -> lambda t
+  val map : ('a -> 'b) -> 'a t -> 'b t
+  val pair : 'a t -> 'b t -> ('a * 'b) t
+  val unit : unit t
+end = struct
+  type 'a dps = tail:bool -> dst:lambda destination -> 'a
+
+  type 'a t = {
+    code : delayed:Constr.t list -> 'a dps;
+    delayed_use_count : int;
+  }
```

Comment:
I think that it is a bit misleading to describe that as an applicative. `map (fun () -> v) unit` and `return v` are not the same. This difference never shows up here anyhow.